File: daisy/processing_queue.py
import collections


class ProcessingQueue:
    """
    A helper class for the scheduler.

    Keeps track of:
    1) blocks that are ready to be scheduled
    2) blocks that are running
    3) how many times blocks have been retried
    """

    def __init__(self, num_roots=0, root_generator=None):
        self.ready_queue = collections.deque()
        self.processing_blocks = set()
        self.block_retries = collections.defaultdict(int)

        self.ready_roots = num_roots
        self.root_generator = root_generator

    @property
    def num_ready(self):
        return self.ready_roots + len(self.ready_queue)

    def get_next(self):
        if self.num_ready > 0:
            if self.ready_roots > 0:
                self.ready_roots -= 1
                return next(self.root_generator)
            else:
                return self.ready_queue.popleft()
        else:
            return None

File: daisy-core/src/processing_queue.rs
use crate::block::Block;
use std::collections::{HashMap, VecDeque};

/// Tracks blocks that are ready, currently processing, and retry counts for a
/// single task.
pub struct ProcessingQueue {
    pub ready_queue: VecDeque<Block>,
    pub processing_blocks: std::collections::HashSet<crate::block::BlockId>,
    pub block_retries: HashMap<crate::block::BlockId, u32>,
    ready_roots: i64,
    root_generator: Option<Box<dyn Iterator<Item = Block> + Send>>,
}

impl ProcessingQueue {
    pub fn new(num_roots: i64, root_generator: Option<Box<dyn Iterator<Item = Block> + Send>>) -> Self {
        Self {
            ready_queue: VecDeque::new(),
            processing_blocks: std::collections::HashSet::new(),
            block_retries: HashMap::new(),
            ready_roots: num_roots,
            root_generator,
        }
    }

    pub fn num_ready(&self) -> i64 {
        self.ready_roots + self.ready_queue.len() as i64
    }

    /// Get the next ready block. Drains from the root generator first, then
    /// falls back to the ready queue.
    pub fn get_next(&mut self) -> Option<Block> {
        if self.num_ready() <= 0 {
            return None;
        }

        if self.ready_roots > 0 {
            self.ready_roots -= 1;
            self.root_generator
                .as_mut()
                .and_then(|iter| iter.next())
        } else {
            self.ready_queue.pop_front()
        }
    }
}
